### ISSUE TITLE
New check: "Must use subscript when assigning associative array" on arr+=( val )

### ISSUE THREAD

**Issue by charles-dyfis-net** [koalaman/shellcheck#768]
#### For new checks and feature suggestions
- [X] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [X] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash
declare -A arr=( )
arr+=( "value" )

```

When run in bash, this yields an error:

    bash4: arr: "value": must use subscript when assigning associative array

#### Here's what shellcheck currently says:

    No issues detected!

#### Here's what I wanted or expected to see:

    += cannot be used with an associative array

**Comment by kurahaupo** [comment]
On Thu, 3 Nov 2016, Charles Duffy wrote:
> #### Here's what I wanted or expected to see:
>
>     += cannot be used with an associative array

But it can:
```
declare -A X=([a]=2 [b]=3 [c]=5)
X+=([d]=6)
declare -p X
```
produces as output:
```
declare -A X='([a]="2" [b]="3" [c]="5" [d]="6" )'
```

**Comment by charles-dyfis-net** [comment]
The description of what to check for was incomplete, then, but the specific code (providing only a value but no key) remains an exemplar of an uncaught error.

**Comment by kurahaupo** [comment]
> the specific code (providing only a value but no key) remains an exemplar of an uncaught error.

Agreed, from a human perspective.

But variables and functions have dynamic scope rather than lexical scope,
which means all sorts of "obvious" stuff isn't half as obvious as you think
it should be.

Usually it's obvious, and shellcheck does do some value-tracking, but in the
most general case you can't tell whether a "declare" statement is
encountered before an array-append statement (unless you actually run (or
emulate) the program), so you cannot tell by inspection whether or not a
variable is an associative array at a given point in the program.

**Comment by timvisher** [comment]
Maybe a warn or a note then?​

**Comment by koalaman** [comment]
ShellCheck's variable typing algorithm is oversimplified and sloppy, but it's there none the less (improvements coming!).

5669702 adds warnings for missing required subscripts, unquoted literals like  `arr=( var=value )` and `array=( [var] = value )` , and empty strings like you'd get in `arr=( [var]= value )`.
